File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/NetworkManager.java
package im.status.ethereum.module;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;
import org.json.JSONException;
import statusgo.Statusgo;
import org.json.JSONObject;

public class NetworkManager extends ReactContextBaseJavaModule {
    private ReactApplicationContext reactContext;
    private Utils utils;

    public NetworkManager(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.utils = new Utils(reactContext);
    }

    @Override
    public String getName() {
        return "NetworkManager";
    }

    @ReactMethod
    public void startSearchForLocalPairingPeers(final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.startSearchForLocalPairingPeers(), callback);
    }

    @ReactMethod
    public void getConnectionStringForBootstrappingAnotherDevice(final String configJSON, final Callback callback) throws JSONException {
        final JSONObject jsonConfig = new JSONObject(configJSON);
        final JSONObject senderConfig = jsonConfig.getJSONObject("senderConfig");
        final String keyUID = senderConfig.getString("keyUID");
        final String keyStorePath = this.utils.getKeyStorePath(keyUID);
        senderConfig.put("keystorePath", keyStorePath);

        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.getConnectionStringForBootstrappingAnotherDevice(jsonConfig.toString()), callback);
    }

    @ReactMethod
    public void inputConnectionStringForBootstrapping(final String connectionString, final String configJSON, final Callback callback) throws JSONException {
        final JSONObject jsonConfig = new JSONObject(configJSON);
        final JSONObject receiverConfig = jsonConfig.getJSONObject("receiverConfig");
        final String keyStorePath = this.utils.pathCombine(this.utils.getNoBackupDirectory(), "/keystore");
        receiverConfig.put("keystorePath", keyStorePath);
        receiverConfig.getJSONObject("nodeConfig").put("rootDataDir", this.utils.getNoBackupDirectory());

        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.inputConnectionStringForBootstrapping(connectionString, jsonConfig.toString()), callback);
    }

    @ReactMethod
    public void sendTransactionWithSignature(final String txArgsJSON, final String signature, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.sendTransactionWithSignature(txArgsJSON, signature), callback);
    }

    @ReactMethod
    public void sendTransaction(final String txArgsJSON, final String password, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.sendTransaction(txArgsJSON, password), callback);
    }

    @ReactMethod
    public void callRPC(final String payload, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.callRPC(payload), callback);
    }

    @ReactMethod
    public void callPrivateRPC(final String payload, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.callPrivateRPC(payload), callback);
    }

    @ReactMethod
    public void recover(final String rpcParams, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.recover(rpcParams), callback);
    }

}

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/NetworkManager.kt
package im.status.ethereum.module

import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import org.json.JSONException
import org.json.JSONObject
import statusgo.Statusgo

class NetworkManager(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private val utils = Utils(reactContext)

    override fun getName() = "NetworkManager"

    @ReactMethod
    fun startSearchForLocalPairingPeers(callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "StartSearchForLocalPairingPeers",
            requestBody = "",
            statusgoFunction = { Statusgo.startSearchForLocalPairingPeers() },
            callback = callback
        )
    }

    @ReactMethod
    fun getConnectionStringForBootstrappingAnotherDevice(configJSON: String, callback: Callback) {
        val jsonConfig = JSONObject(configJSON)
        val senderConfig = jsonConfig.getJSONObject("senderConfig")
        val keyUID = senderConfig.getString("keyUID")
        val keyStorePath = utils.getKeyStorePath(keyUID)
        senderConfig.put("keystorePath", keyStorePath)
        val jsonString = jsonConfig.toString()

        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "GetConnectionStringForBootstrappingAnotherDevice",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.getConnectionStringForBootstrappingAnotherDevice(jsonString) },
            callback
        )
    }

    @ReactMethod
    fun inputConnectionStringForBootstrapping(connectionString: String, configJSON: String, callback: Callback) {
        val receiverClientConfig = JSONObject(configJSON)
        val params = JSONObject().apply {
            put("connectionString", connectionString)
            put("receiverClientConfig", receiverClientConfig)
        }
        val jsonString = params.toString()
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "InputConnectionStringForBootstrappingV2",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.inputConnectionStringForBootstrappingV2(jsonString) },
            callback
        )
    }

    @ReactMethod
    fun sendTransactionWithSignature(txArgsJSON: String, signature: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "SendTransactionWithSignature",
            requestBody = txArgsJSON,
            statusgoFunction = { Statusgo.sendTransactionWithSignature(txArgsJSON, signature) },
            callback
        )
    }

    @ReactMethod
    fun sendTransaction(txArgsJSON: String, password: String, callback: Callback) {
        val jsonParams = JSONObject().apply {
            put("txArgs", JSONObject(txArgsJSON))
            put("password", password)
        }
        val jsonString = jsonParams.toString()
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "SendTransactionV2",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.sendTransactionV2(jsonString) },
            callback
        )
    }

    @ReactMethod
    fun callRPC(payload: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "CallRPC",
            requestBody = payload,
            statusgoFunction = { Statusgo.callRPC(payload) },
            callback = callback
        )
    }

    @ReactMethod
    fun callPrivateRPC(payload: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "CallPrivateRPC",
            requestBody = payload,
            statusgoFunction = { Statusgo.callPrivateRPC(payload) },
            callback = callback
        )
    }

    @ReactMethod
    fun recover(rpcParams: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "Recover",
            requestBody = rpcParams,
            statusgoFunction = { Statusgo.recover(rpcParams) },
            callback
        )
    }

    @ReactMethod
    fun getConnectionStringForExportingKeypairsKeystores(configJSON: String, callback: Callback) {
        val jsonConfig = JSONObject(configJSON)
        val senderConfig = jsonConfig.getJSONObject("senderConfig")
        val keyUID = senderConfig.getString("loggedInKeyUid")
        val keyStorePath = utils.getKeyStorePath(keyUID)
        senderConfig.put("keystorePath", keyStorePath)
        val jsonString = jsonConfig.toString()

        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "GetConnectionStringForExportingKeypairsKeystores",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.getConnectionStringForExportingKeypairsKeystores(jsonString) },
            callback = callback
        )
    }

    @ReactMethod
    fun inputConnectionStringForImportingKeypairsKeystores(connectionString: String, configJSON: String, callback: Callback) {
        val keystoreFilesReceiverClientConfig = JSONObject(configJSON)
        val receiverConfig = keystoreFilesReceiverClientConfig.getJSONObject("receiverConfig")
        val keyStorePath = utils.pathCombine(utils.getNoBackupDirectory(), "/keystore")
        receiverConfig.put("keystorePath", keyStorePath)

        val params = JSONObject().apply {
            put("connectionString", connectionString)
            put("keystoreFilesReceiverClientConfig", keystoreFilesReceiverClientConfig)
        }
        val jsonString = params.toString()
        StatusBackendClient.executeStatusGoRequestWithCallback(
            endpoint = "InputConnectionStringForImportingKeypairsKeystoresV2",
            requestBody = jsonString,
            statusgoFunction = { Statusgo.inputConnectionStringForImportingKeypairsKeystoresV2(jsonString) },
            callback = callback
        )
    }
}
